Создать стажировку

NadirCianid opened the pull request.

InnMish committed:
fix

NadirCianid committed:
Начал разработку добавления стажировки

NadirCianid committed:
Добавил действие "Создать стажировку" (без тестов)

nn1ck0 commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,67 @@
+    private final InternshipValidation internshipValidation;
+
+    public InternshipDto create(InternshipDto internshipDto) {
+        validationDto(internshipDto);
```
Лучшше вынести валидацию в отдельный класс, а то контроллер уже странно смотрится

nn1ck0 commented on src/main/java/faang/school/projectservice/validation/InternshipValidation.java:
```diff
@@ -0,0 +1,49 @@
+
+@Component
+@RequiredArgsConstructor
+public class InternshipValidation {
```
Я думаю лучше сделать методы, которые будут бросать исключения, заинжектить этот класс в сервис и уже использовать его методы и передавать нужные параметры, чем инжектить все зависимости в валидацию и использовать приватные методы

NadirCianid replied:
Пока не вижу причин это делать

NadirCianid committed:
Вынес валидацию из контроллера и начал писать тесты

NadirCianid committed:
инкапсулировал валидаторы

NadirCianid committed:
Добавил действие обновления стажировки

RickHammerson commented on src/main/java/faang/school/projectservice/controller/internship/InternshipControllerValidation.java:
```diff
@@ -0,0 +1,52 @@
+        if (internshipDto.getMentorId() == null) {
+            throw new DataValidationException(EMPTY_INTERNSHIP_MENTOR_ID_EXCEPTION.getMessage());
+        }
+        if (internshipDto.getInternsIds() == null || internshipDto.getInternsIds().size() == 0) {
```
Рекоммендую использовать метод от спринга для проверки коллекции CollectionUtils.isEmpty(internshipDto.getInternsIds()). Делает всё то же самое, но код чуть короче)

NadirCianid replied:
Спасибо!

RickHammerson commented on src/main/java/faang/school/projectservice/service/internship/InternshipService.java:
```diff
@@ -0,0 +1,106 @@
+    private void processCompletedInternship(Internship internship) {
+        List<TeamMember> interns = internship.getInterns();
+        var internsToBeHired = interns.stream()
+                .filter(intern -> {
```
Метод всегда должен быть как можно меньше, чтобы не загружать глаза большим количеством строк
Я бы вынес эту большую логику фильтра в отдельный метод, который возвращает предикат.

NadirCianid replied:
Тоже об этом подумал

RickHammerson commented on src/main/java/faang/school/projectservice/service/internship/InternshipService.java:
```diff
@@ -0,0 +1,106 @@
+                throw new RoleProcessingException(ABSENT_INTERN_ROLE_EXCEPTION.getMessage());
+            }
+
+            if (internRoles.size() == 1) {
```
А если несколько ролей будет? Допустим, кто-то после тебя код изменит и их будет больше одного

NadirCianid replied:
Согласен, не в том месте удалял роль стажера

NadirCianid committed:
Добавил действие "Получить стажировки проекта с фильтрами"

NadirCianid committed:
Добавил действия "Получить все стажировки" и "Получить стажировку по Id"

RickHammerson commented on src/main/java/faang/school/projectservice/service/internship/filter/InternshipEndDateFilter.java:
```diff
@@ -0,0 +1,29 @@
+        return internships.stream()
+                .filter(internship -> {
+                    LocalDateTime internshipEndDate = internship.getEndDate();
+                    LocalDateTime endDatePattern = filters.getEndDatePattern();
```
По сути , endDatePattern можно её один раз достать и использовать. Мне кажется,лучше вынести из стрима

RickHammerson commented on src/main/java/faang/school/projectservice/service/internship/filter/InternshipStartDateFilter.java:
```diff
@@ -0,0 +1,29 @@
+        return internships.stream()
+                .filter(internship -> {
+                    LocalDateTime internshipStartDate = internship.getStartDate();
+                    LocalDateTime startDatePattern = filters.getStartDatePattern();
```
То же самое,как в endDatePattern

RickHammerson commented on src/main/java/faang/school/projectservice/controller/internship/InternshipController.java:
```diff
@@ -37,4 +37,12 @@ public InternshipDto update(InternshipDto internshipDto) {
+        return internshipService.getAllInternships();
+    }
+
+    public InternshipDto getInternshipById(long internshipId) {
```
Лучше принимать Long в качестве параметра, т.к если просто дернуть эндпоинт без передачи id, у тебя будет по дефолту 0L, а не null, и в метод что-то передастся

RickHammerson commented on src/main/java/faang/school/projectservice/service/internship/InternshipService.java:
```diff
@@ -65,65 +58,14 @@ public List<InternshipDto> getInternshipsOfProject(long projectId, InternshipFil
-        internship.setInterns(interns);
+    public InternshipDto getInternshipById(long internshipId) {
+        Internship internship = internshipRepository.findById(internshipId)
+                .orElseThrow(() -> new DataValidationException(NON_EXISTING_INTERNSHIP_EXCEPTION.getMessage()));
```
Насчет переносов не уверен, будет ли так лучше обрывать
Лучше спросить у Сабины. Возможно, можно ввести на проекте код стайл

NadirCianid replied:
@Sabina-Vishnya, как в данном случае быть с переносом? Норм или нет?

Sabina-Vishnya replied:
Я за то, чтобы код помещался в окно идеи, в данном случае перенос к месту. 
Еще видела, что некоторые прям по точкам просто переносят, но мне такой вариант вполне понятен тоже. 
В идеи рамка есть и выходя за рамки сложнее воспринимать уже остальное, ты как бы фокусируешься, когда код у тебя попадает в одно поле видимости. 
Есть даже некий стандарт https://www.oracle.com/java/technologies/javase/codeconventions-indentation.html
не только у джавы такое, но вот исторически сложилась так, что настроены идеи на 80 символов обычно. Кажется это из-за того, что печатали исходники и удобнее было. Щас уже некоторым всё равно и это не требуется строго соблюдать прям, изменились времена, но по холивару на эту тему на работе есть мнение, что многим всё же такой вариант понятнее по сей день) Можно конечно растянуть всё, но в большинстве случаев удобнее не растягивать при чтении. Так что, я за такую, если можно назвать, компактность)
Радмир, про код стайл хорошо подсветил, была мысль такое предложить, возможно надо бы, только кто готов себе туду завести?) Можем на дейли обсудить еще, у нас там есть разные задачки, эту мб тоже добавим

RickHammerson commented on src/main/java/faang/school/projectservice/service/internship/InternshipServiceUtility.java:
```diff
@@ -0,0 +1,88 @@
+        assignRolesToInterns(internsToBeHired);
+    }
+
+    private Predicate<TeamMember> checkCompletionOfTasks() {
```
Из видосов Влада помню, что лучше соблюдать очередность полей и методов по модификатору доступа
Так что лучше сделать public методы выше private и других

RickHammerson commented on src/main/java/faang/school/projectservice/service/internship/InternshipServiceUtility.java:
```diff
@@ -0,0 +1,88 @@
+
+@Component
+@RequiredArgsConstructor
+class InternshipServiceUtility {
```
Не уверен, что стоит разделять сервис на сервис и утилиты
Надо у Сабины уточнить. А то как будто один класс на 2 растащили

NadirCianid replied:
Растащили, чтобы нормально протестировать код в приватных методах. + так сервис лучше читается, как мне кажется

Sabina-Vishnya replied:
"нормально протестировать код в приватных методах" - не очень поняла про что это?
Учитывая, что не все тесты есть - это скорее всего доделаешь после основных правок? 

Сложно понять, что происходит, поскольку глобально разницы сервиса, валидатора и утилиты нету. Сервис лучше читается тебе, потому что ты больше остальных погружен в происходящее и бизнес-логику текущую. Спустя время наверняка тебе захочется этот код переделать.
Метод toEntity в утилитах, который и мапит еще другой toEntity, и в репозитории идет. Суть метода в чем? Название явно ничего не говорит, а лишь вводит в заблуждение. 
Вадлидатор и сервис используют одни и те же методы для работы с бд. Класс утилит также проверки имеет. В чем смысл такого растаскивания, когда тут не прослеживается разделение ответственности?

NadirCianid committed:
Мелкие правки:
- Вынес объявление datePattern в фильтрах из стрима;
- Поправил порядок методов в InternshipServiceUtility;
- Поменял тип параметра internshipId в контроллере на обертку (Long);

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/controller/internship/InternshipController.java:
```diff
@@ -0,0 +1,52 @@
+    public InternshipDto update(InternshipDto internshipDto) {
+        internshipControllerValidation.validationDto(internshipDto);
+
+        if (internshipDto.getId() == null) {
```
эта проверка и ниже идентичны почти, там лонги в обоих случаях, а сообщение аргументом можно передать и более того, в твоем валидаторе InternshipControllerValidation есть также такие проверки, которые проверяют лонг на null. Не привязывайся к объекту там, где проверяешь только поля, вот в данном случае у тебя много частей с повторением на проверки полей конкретного типа, хоть это и может показаться неочевидным, но тут явно можно валидацию уменьшить и переиспользовать

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/controller/internship/InternshipControllerValidation.java:
```diff
@@ -0,0 +1,53 @@
+
+        LocalDateTime startDate = internshipDto.getStartDate();
+        LocalDateTime endDate = internshipDto.getEndDate();
+        if (startDate.isBefore(endDate.minus(3, ChronoUnit.MONTHS))) {
```
что за 3 неизвестное никому?) это в константу стоит вынести

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipServiceValidation.java:
```diff
@@ -0,0 +1,68 @@
+
+@Component
+@RequiredArgsConstructor
+class InternshipServiceValidation {
```
VacancyRepository нигде не используется тут

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,31 @@
+    @Mapping(target = "project", ignore = true)
+    @Mapping(target = "mentorId", ignore = true)
+    @Mapping(target = "interns", ignore = true)
+    Internship toEntity(InternshipDto dto);
```
норм ли нам игнорить поля эти? а то что-то не первый раз вижу игнор полей, не нахватаем ли мы багов и кривую логику из-за этого?

NadirCianid replied:
Мы их игнорируем, потому что не можем в мапере обратиться в репозиторий, чтобы их установить

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipServiceUtility.java:
```diff
@@ -0,0 +1,88 @@
+
+            if (internRoles.size() == 1) {
+                internRoles.add(TeamRole.DEVELOPER);
+                //TODO: нормально ли делать по умолчанию бывших стажеров разработчиками, если до этого у них не было другой роли?
```
Если требуется доп. разграничение ролей и прав доступа, то присвоение роли разработчика по дефолту может не подойти. У нас вроде такого пока нет, поэтому кажется норм, но непонятно у стажеров роль интерна была или нет? у бывших стажеров должна быть удалена роль интерна и присвоена какая-то другая ведь, т.е. стажер-разработчик после удачной стажировки будет с ролью разработчика, а роль интерна у него удалена, как это "не было другой роли"?

NadirCianid replied:
У стажеров роль интерна должна быть в любом случае
Также у них в теории должна быть и вторая роль, указывающая их специальность

NadirCianid replied:
"не было другой роли", кроме стажера

NadirCianid committed:
Вынес проверки объектов на null в отдельный метод, чтобы уменьшить дублирование

NadirCianid committed:
Мелкие правки:
- Вынес в константу длительность стажировки;
- Удалил неиспользуемое поле в InternshipServiceValidation

NadirCianid committed:
Вернул валидацию стажировок в сервис, но вынес взаимодействия с TeamMemberRepository & ProjectRepository в соответсвующее сервисы

NadirCianid committed:
Расформировал класс InternshipServiceUtility:
- Вынес найм интернов в TeamMemberService;
- Вынес заполнение сущности стажировки в класс InternshipFiller;
- Разбил валидацию создания/обновления в сервисе на одноцелевые методы.

NadirCianid committed:
- Исправил заполнение сущностей стажировок в сохранении и обновлении;
- Обновил код под REST

NadirCianid committed:
Исправил объявление метода получения стажировок с фильтрами (заменил метод HTTP на POST)

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipFiller.java:
```diff
@@ -0,0 +1,29 @@
+@Component
+@RequiredArgsConstructor
+public class InternshipFiller {
+    private final InternshipMapper internshipMapper;
```
не вижу тут использования маппера этого

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipFiller.java:
```diff
@@ -0,0 +1,29 @@
+
+@Component
+@RequiredArgsConstructor
+public class InternshipFiller {
```
~public~

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipFiller.java:
```diff
@@ -0,0 +1,29 @@
+        internship.setProject(projectRepository.getProjectById(projectId));
+        internship.setMentorId(teamMemberRepository.findById(mentorId));
+
+        var interns = internIds.stream()
```
у jpa репозитория кажется есть метод findAllById
посмотри, вдруг, он подойдет сюда и тогда оптимальнее и красивее выйдет код

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipService.java:
```diff
@@ -0,0 +1,126 @@
+
+
+    public InternshipDto create(InternshipDto internshipDto) {
+        verifyMentorsProject(internshipDto);
```
подумай над рефакторингом, у тебя сейчас снова дублирование по сути, например:
1 метод верификации, внутри вызывает getProjectById
2 метод getTeamMembersProject фактически тоже, тк внутри вызывает findById
эти же методы у тебя повторно вызываются для энтити в InternshipFiller

Аналогично и для update метода

Попробуй оптимизировать, код у тебя вероятнее всего упростится

NadirCianid committed:
Оптимизировал код в проверках InternshipService

NadirCianid committed:
Удалил неиспользуемый метод из ProjectService

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipService.java:
```diff
@@ -0,0 +1,125 @@
+
+
+    public InternshipDto create(InternshipDto internshipDto) {
+        verifyMentorsProject(internshipDto.getProjectId(), internshipDto.getMentorId());
```
дублируешь логику в валидаторах и методах, вызовы те же
и у тебя сервис энтити возвращает, какой тогда смысл в вызовах таких сервисов?
почему нельзя тогда сделать проще, например, вызов трех репозиториев,
типа:
```
Project project = projectRepository.getProjectById(internshipDto.getProjectId());
TeamMember teamMember = teamMemberRepository.findById(internshipDto.getMentorId());
List<TeamMember> teamMemberList = teamMemberRepository.findAllByIds(internshipDto.getInternsIds());
```
В каждом есть проверка, кроме последнего вызова, где список. Эти же данные потом сетятся в филлере. Суть та же, те же энтити и другие репозитории, почему не упростить тогда до похожего примитивного вызова с проверкой? 

подумай о подобном, чтобы не дублироваться
и у тебя какие-то сервисы возвращают дто, а какие-то энтити, например TeamMemberService, почему так - непонятно,. Тут бы тоже привести в порядок.

Это, если что, лишь часть метода для примера, который можно было бы поправить. Тебе стоит начать с того, чтобы просмотреть подобные вещи в других частях кода, а они есть и поправить их. Ты ведь таким образом и код впоследствии можешь упрощать, а сейчас усложняешь, накручивая проверки, которые у тебя в вызовах и так есть.

NadirCianid committed:
Убрал дублирование кода в InternshipService

NadirCianid committed:
Сменил эндпоинт метода getInternshipsOfProject

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipFiller.java:
```diff
@@ -0,0 +1,24 @@
+
+@Component
+@RequiredArgsConstructor
+class InternshipFiller {
```
Бины тут не используются + от этого класса можно отказаться в виде филлера и сетить в сервисе тогда уже, а этот класс лучше сделать валидатором. Просто
все свои методы verify перетащить сюда и переименовать класс - всё. 
Больше здесь ничего не будет, кроме твоих приватных методов верифаев из сервиса, которым ты модификатор доступа только поправишь.

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/service/internship/InternshipService.java:
```diff
@@ -0,0 +1,119 @@
+        verifyExistenceOfAllInterns(interns, internshipDto.getInternsIds().size());
+
+        Internship internship = internshipMapper.toEntity(internshipDto);
+        internshipFiller.fillEntity(internship, project, mentor, interns);
```
- Про филлер тебе написала, сеттеры сюда притаскивай, а три метода verify утаскивай

- А вот эта часть
```
Project project = projectRepository.getProjectById(internshipDto.getProjectId());
TeamMember mentor = teamMemberRepository.findById(internshipDto.getMentorId());
List<TeamMember> interns = teamMemberRepository.findAllByIds(internshipDto.getInternsIds());

verifyMentorsProject(project, mentor);
verifyExistenceOfAllInterns(interns, internshipDto.getInternsIds().size());
```
дублируется в create и update методах. Можно попробовать выделить в общий метод. 
Например, с помощью приватного класса хранить `Project, TeamMember, List<TeamMember>`. А часть повторяющегося кода вынести в метод и возвращать объект с нужными данными.

Sabina-Vishnya commented on src/main/java/faang/school/projectservice/mapper/InternshipMapper.java:
```diff
@@ -0,0 +1,37 @@
+    @Mapping(target = "project", ignore = true)
+    @Mapping(target = "mentorId", ignore = true)
+    @Mapping(target = "interns", ignore = true)
+    void update(InternshipDto dto, @MappingTarget Internship entity);
```
Лучше переименовать, иначе будет сбивать столку, что это часть маппинга.
Можно fillEntityForUpdate или что-то в этом духе, просто название понятнее метода

NadirCianid replied:
Это же буквально часть маппинга. Просто в этом методе не создается новая сущность, а обновляется существующая.

NadirCianid committed:
Исправление замечаний из ревью:
- Вернул fillEntity в сервис;
- Вынес верификацию данных стажировки из сервиса в отдельный класс;
- Выделили дублирующуюся логику в сервисе в отдельный метод с использованием класса InternshipData в качестве контейнера данных стажировки.

NadirCianid committed:
Написал тесты
